modules/market/src/main/java/com/opengamma/strata/market/curve/CurveGroupDefinition.java

```diff
@@ -100,27 +119,57 @@ public static CurveGroupDefinition of(CurveGroupName name, Collection<CurveGroup
    *
    * @param name  the name of the curve group
    * @param entries  details of the curves in the group
+   * @param curveDefinitions  definitions which specify how the curves are calibrated
    */
   @ImmutableConstructor
-  CurveGroupDefinition(CurveGroupName name, Collection<CurveGroupEntry> entries) {
+  CurveGroupDefinition(
+      CurveGroupName name,
+      Collection<CurveGroupEntry> entries,
+      Collection<NodalCurveDefinition> curveDefinitions) {
+
     this.name = ArgChecker.notNull(name, "name");
     this.entries = ImmutableList.copyOf(entries);
-    entriesByName = entries.stream().collect(toImmutableMap(entry -> entry.getCurveDefinition().getName(), entry -> entry));
+    this.curveDefinitions = ImmutableList.copyOf(curveDefinitions);
+    entriesByName = entries.stream().collect(toImmutableMap(entry -> entry.getCurveName(), entry -> entry));
+    curveDefinitionsByName = curveDefinitions.stream().collect(toImmutableMap(def -> def.getName(), def -> def));
+    validate();
+  }
+
+  @ImmutableValidator
+  private void validate() {
+    Set<CurveName> missingEntries = Sets.difference(curveDefinitionsByName.keySet(), entriesByName.keySet());
+
+    if (!missingEntries.isEmpty()) {
+      throw new IllegalArgumentException("An entry must be provided for every curve definition but the following " +
+          "curves have a definition but no entry: " + missingEntries);
+    }
   }
 
   //-------------------------------------------------------------------------
   /**
-   * Finds the entry for the curve group with the specified name.
+   * Finds the entry for the curve with the specified name.
    * <p>
-   * If the group is not found, optional empty is returned.
+   * If the curve is not found, optional empty is returned.
    *
    * @param curveName  the name of the curve
-   * @return the entry for the curve group with the specified name
+   * @return the entry for the curve with the specified name
    */
   public Optional<CurveGroupEntry> findEntry(CurveName curveName) {
     return Optional.ofNullable(entriesByName.get(curveName));
   }
 
+  /**
+   * Finds the definition for the curve with the specified name.
+   * <p>
+   * If the curve is not found, optional empty is returned.
+   *
+   * @param curveName  the name of the curve
+   * @return the definition for the curve with the specified name
+   */
+  public Optional<NodalCurveDefinition> findDefinition(CurveName curveName) {
```

Comment:
Field is `curveDefinitions` but this method is just `definition`. Should stick to one name.